CASSANDRA.md

```diff
@@ -0,0 +1,543 @@
+      ).to[List]
+    }
+  }
+
+  val getAllByCountry: Country => List[WeatherStation] = WeatherStation.getAllByCountry(cache, session)_
+
+  getAllByCountry(Country("UK"))
+
+  session.close()
+  cluster.close()
+}
+```
+
+We need to create a new `TypeCodec` and register it in the `CodecRegistry`.
+
+**Phantom (v1.22.0)**
+```scala
+import com.websudos.phantom.builder.primitives.Primitive
+import com.websudos.phantom.builder.query.CQLQuery
+import com.websudos.phantom.builder.syntax.CQLSyntax
+import com.websudos.phantom.connectors.RootConnector
+import com.websudos.phantom.db._
+import com.websudos.phantom.dsl._
+
+import scala.concurrent.Future
+import scala.util.Try
+
+object Phantom extends App {
+
+  case class Country(code: String) extends AnyVal
+
+  object Country {
+
+    implicit object CountryIsPrimitive extends Primitive[Country] {
+
+      override type PrimitiveType = Country
+
+      override def fromRow(column: String, row: Row): Try[Country] =
+        nullCheck(column, row) {
+          r => Country(r.getString(column))
+        }
+
+      override val cassandraType: String = CQLSyntax.Types.Text
+
+      override def fromString(code: String): Country = Country(code)
+
+      override def asCql(country: Country): String = CQLQuery.empty.singleQuote(country.code)
+
+      override val clz: Class[CountryIsPrimitive.PrimitiveType] = classOf[Country]
+    }
+
+    type Column[Owner <: CassandraTable[Owner, Record], Record] = PrimitiveColumn[Owner, Record, Country]
+  }
+
+  case class WeatherStation(country: Country, city: String, stationId: String, entry: Int, value: Int)
+
+  case class People(name: String, age: Int)
+
+  abstract class WeatherStationCF(override val tableName: String) extends CassandraTable[WeatherStationCF, WeatherStation] with RootConnector {
+
+    object country extends Country.Column(this) with PartitionKey[Country]
+    object city extends StringColumn(this) with PrimaryKey[String]
+    object stationId extends StringColumn(this) with PrimaryKey[String] {
+      override lazy val name: String = "station_id"
+    }
+    object entry extends IntColumn(this) with PrimaryKey[Int]
+    object value extends IntColumn(this)
+
+    override def fromRow(r: Row): WeatherStation =
+      WeatherStation(country(r), city(r), stationId(r), entry(r), value(r))
+  }
+
+  abstract class WeatherStationQueries extends WeatherStationCF("weather_station") {
+
+    def getAllByCountry(country: Country): Future[List[WeatherStation]] =
+      select.where(_.country eqs country).fetch()
+  }
+
+  class DB(ks: KeySpaceDef) extends DatabaseImpl(ks) {
+
+    object stations extends WeatherStationQueries with connector.Connector
+  }
+
+  val db = new DB(ContactPoint.local.keySpace("db"))
+
+  val result = db.stations.getAllByCountry(Country("UK"))
+
+  result.onComplete(_ => db.shutdown())
+}
+```
+
+We need to define a new `Column` type and use it while defining the data model.
```

Comment:
Phantom requires definition of a new `Column` type to be used when defining the data model.